src/crypto/NoteEncryption.cpp

```diff
@@ -0,0 +1,161 @@
+    key[31] |= 64;
+}
+
+void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &seed,
+    unsigned char nonce
+   )
+{
+    unsigned char K_tmp[64];
+    CSHA512 hasher;
+
+    // Trim the first byte from seed
+    std::vector<unsigned char> first_248(seed.begin() + 1, seed.end());
+
+    if (nonce >= 2) {
+        throw std::logic_error("KDF nonce space exceeded");
+    }
+
+    // First bit is the 'nonce' bit or `i`
+    // Second bit is zero
+    first_248[0] &= 0xFC;
+    first_248[0] |= nonce;
+
+    hasher.Write(&first_248[0], 31);
+    hasher.Write(dhsecret.begin(), crypto_scalarmult_BYTES);
+    hasher.Write(epk.begin(), crypto_scalarmult_BYTES);
+    hasher.Write(pk_enc.begin(), crypto_scalarmult_BYTES);
+    hasher.Finalize(K_tmp);
+
+    memcpy(K, K_tmp + 32, 32);
+}
+
+template<size_t MLEN>
+NoteEncryption<MLEN>::NoteEncryption(uint256 seed) : nonce(0), seed(seed) {
+    // All of this code assumes crypto_scalarmult_BYTES is 32
+    // There's no reason that will _ever_ change, but for
+    // completeness purposes, let's check anyway.
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_BYTES);
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_SCALARBYTES);
+    BOOST_STATIC_ASSERT(NOTEENCRYPTION_AUTH_BYTES == crypto_aead_chacha20poly1305_ABYTES);
+
+    // Create the ephemeral keypair
+    randombytes_buf(esk.begin(), crypto_scalarmult_SCALARBYTES);
+    epk = generate_pubkey(esk);
+}
+
+template<size_t MLEN>
+NoteEncryption<MLEN>::~NoteEncryption() {
+    sodium_memzero(epk.begin(), crypto_scalarmult_BYTES);
+    sodium_memzero(esk.begin(), crypto_scalarmult_SCALARBYTES);
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Ciphertext NoteEncryption<MLEN>::encrypt
+                                          (const uint256 &pk_enc,
+                                           const NoteEncryption<MLEN>::Plaintext &message
+                                          )
+{
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), esk.begin(), pk_enc.begin()) != 0) {
+        throw std::logic_error("Could not create DH secret");
+    }
+
+    // Construct the symmetric key
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF(K, dhsecret, epk, pk_enc, seed, nonce);
+
+    // Increment the number of encryptions we've performed
+    nonce++;
+
+    // The nonce is zero for our purposes
+    unsigned char cipher_nonce[crypto_aead_chacha20poly1305_NPUBBYTES] = {};
+
+    NoteEncryption<MLEN>::Ciphertext ciphertext;
+
+    crypto_aead_chacha20poly1305_encrypt(ciphertext.begin(), NULL,
+                                         message.begin(), MLEN,
+                                         NULL, 0, // no "additional data"
+                                         NULL, cipher_nonce, K);
+
+    return ciphertext;
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Plaintext NoteEncryption<MLEN>::decrypt
```

Comment:
Yes, I plan on doing this, but for a different reason. We will probably be attempting to decrypt a lot of ciphertexts, so it will be useful to provide `sk_enc` *once* from which we can derive `pk_enc` once for the following KDFs, an optimization not present in the current code.